### PR TITLE
Clear modem phone book on serial ports shutdown

### PR DESCRIPTION
**Pull request by NicknineTheEagle**
This fixes a memory leak of sorts introduced in https://github.com/dosbox-staging/dosbox-staging/pull/226.
Turns out serial ports are re-initialized every time you change any of serialX parameters at runtime and this, in turn, re-parses phone book file, potentially filling up the phones list with endless duplicate entries.

### CODE REVIEW

**Comment by NicknineTheEagle** [comment]
> 
> 
> > Oh, right, I could have sworn it was uppercase, one minute.
> 
> (It's not a callback though, right? If not, then nix the prefix)

I don't get it, what does callback have to do with a prefix? I have gave it a prefix to classify it since it's a global function.

**Comment by kcgen** [comment]
Not sure where previous comments went. 

I think in this case a capitalized MODEM prefix makes sense (and is consistent with other similar functions), given it's called up in the more common serialports code.

**Comment by NicknineTheEagle** [comment]
> 
> 
> Not sure where previous comments went.
> 
> I think in this case a capitalized MODEM prefix makes sense (and is consistent with other similar functions), given it's called up in the more common serialports code.

I force-pushed and previous comments were attached to the old commit so they got left behind.

**Review by kcgen — APPROVED** [review]
Looks good @NicknineTheEagle; over to @dreamer.

File: src/hardware/serialport/softmodem.cpp [hunk 1]
```diff
@@ -96,6 +96,15 @@ bool MODEM_ReadPhonebook(const std::string &filename) {
 	return true;
 }
 
+void MODEM_ClearPhonebook()
+{
+	for (auto entry : phones) {
+		delete entry;
+	}
```
Comment (dreamer) [inline src/hardware/serialport/softmodem.cpp:103]: *Two suggestions, both are completely optional and are not blocking this PR from being merged.*

(Optional improvement)

This loop would be unnecessary if type of `phones` was changed from `std::vector<PhonebookEntry *>` to `std::vector<PhonebookEntry>`. This way we could also do a tiny speedup in softmodem.cpp:92 replacing

```C++
PhoneBookEntry *pbEntry = new PhonebookEntry(phone, address);
phones.push_back(pbEntry);
```
with
```
phones.emplace_back(phone, address);
```

(General comment about using auto in for-each loop in C++)

Good rule of thumb to follow is:

```C++
// xs is a container of raw pointers
for (auto *x : xs) {} // good - gives reader a hint, that xs is full of pointers
for (auto &x : xs) {} // bad - hopefully compiler can optimize it out
for (auto x : xs) {}  // works, but bad - might result in silent performance regression if container changes type during code refactor

// ys is a container of objects
for (auto *y : ys) {} // won't compile
for (auto &y : ys) {} // good
for (auto y : ys) {}  // works, but bad - will call copy constructor passing every object - use it only when it's exactly what you want

// zs is a container of simple types (e.g. ints)
for (auto *z : zs) {} // won't compile
for (auto &z : zs) {} // ok - if you want to modify the container
for (auto z : zs) {}  // ok - preferable, if you don't need to modify container
```
In other words, when dealing with container with pointers, prefer `for (auto *x : xs) …` or (not in this case) `for (const auto *x : xs) …`.

**Review by dreamer — CHANGES_REQUESTED** [review]

File: src/hardware/serialport/softmodem.cpp (hunk 1 above)
Comment (NicknineTheEagle) [reply src/hardware/serialport/softmodem.cpp:103]: I've originally used pointers because push_back would call a copy constructor which is slow but emplace_back seems like it solves this.

File: src/hardware/serialport/serialport.cpp [hunk 2]
```diff
@@ -1254,6 +1254,7 @@ class SERIALPORTS:public Module_base {
 
 #if C_MODEM
 		const Prop_path *pbFilename = section->Get_path("phonebookfile");
+		MODEM_ClearPhonebook();
```
[the remark this answers is not in the document]
Comment (NicknineTheEagle) [reply src/hardware/serialport/serialport.cpp:1257]: Funny, I placed it there in my patch for upstream and then forgot about it.

File: src/hardware/serialport/serialport.cpp [hunk 3]
```diff
@@ -1308,6 +1308,8 @@ class SERIALPORTS:public Module_base {
 	}
 
 	~SERIALPORTS () {
+		MODEM_ClearPhonebook();
```
Comment (dreamer) [inline src/hardware/serialport/serialport.cpp:1311]: Complete nitpick: pedantically speaking, this line should be placed after the `for` loop (so we keep the reverse order compared to initialization).

However, in this case it doesn't matter, `serialports` does not depend on data cleaned in `MODEM_ClearPhonebook()` at all, and likely never will.

**Review by dreamer — APPROVED** [review]
Nice, thanks @NicknineTheEagle! :)

File: src/hardware/serialport/serialport.cpp (hunk 3 above)
Comment (kcgen) [reply src/hardware/serialport/serialport.cpp:1311]: And I suppose if @NicknineTheEagle moves from using pointers to objects like you've nicely suggested @dreamer, we can let the phonebook container clean itself up.

**Comment by kcgen** [comment]
Build fails under aarch64:

```
serialport.cpp:1311:3: error: use of undeclared identifier 
'MODEM_ClearPhonebook'                                               MODEM_ClearPhonebook();                                 
^                                       
42 warnings and 1 error generated.                      
make[4]: *** [Makefile:398: serialport.o] Error 1       make[3]: *** [Makefile:540: all-recursive] Error 1      make[2]: *** [Makefile:484: all-recursive] Error 1      make[1]: *** [Makefile:401: all-recursive] Error 1      
make: *** [Makefile:342: all] Error 2

$ clang -v
clang version 10.0.1 (https://github.com/termux/Packages 2f1431d824f449f09ef50aa365cdd7ed5b8c159d)       Target: aarch64-unknown-linux-android
Thread model: posix
InstalledDir: /data/data/com.termux/files/usr/bin

$ uname -a
Linux localhost 4.4.153-18800027 #2 SMP PREEMPT Tue Jun 30 22:33:08 KST 2020 aarch64 Android
```

Maybe missing its prototype (or not visible in serialport.cpp)

**Comment by dreamer** [comment]
I don't see what could be wrong… our daily scheduled ARMv8 build should arrive in ~1h, let's see if it fails as well.
